File: JS/cedrus/src/models/PropertyModel.ts
import { isDate, isNullOrUndefined, isString } from "@juniper-lib/util";
import { DataType, DataTypeMap, StorageType, isDataType, isStorageType } from "./DataType";
import { ENTITY, EntityModel } from "./EntityModel";
import { ICreationTracked, isICreationTracked } from "./ICreationTracked";
import { ISequenced, isISequenced } from "./ISequenced";
import { ITypeStamped, isTypeStamped } from "./ITypeStamped";
import { TypedNameOrId } from "./NameOrId";
import { PROPERTY_TYPE, PropertyTypeModel, isPropertyTypeModel } from "./PropertyTypeModel";
import { isUserModel, UserModel } from "./UserModel";

export interface PropertyModel<DataTypeT extends DataType = DataType, StorageTypeT extends StorageType = StorageType>
    extends ISequenced,
    ITypeStamped<"property">,
    ICreationTracked {
    type: PropertyTypeModel<DataTypeT, StorageTypeT>;
    value: DataTypeMap[DataTypeT][StorageTypeT];
    units: string;
    reference?: EntityModel;
    entity?: EntityModel;
    updatedBy: UserModel;
    updatedOn: Date;
}

export function isPropertyModel(obj: unknown): obj is PropertyModel<DataType, StorageType>;
export function isPropertyModel<DataTypeT extends DataType>(obj: unknown, dataType: DataTypeT): obj is PropertyModel<DataTypeT, StorageType>;
export function isPropertyModel<StorageTypeT extends StorageType>(obj: unknown, storageType: StorageTypeT): obj is PropertyModel<DataType, StorageTypeT>;
export function isPropertyModel<DataTypeT extends DataType, StorageTypeT extends StorageType>(obj: unknown, dataType: DataTypeT, storageType: StorageTypeT): obj is PropertyModel<DataTypeT, StorageTypeT>;
export function isPropertyModel<DataTypeT extends DataType, StorageTypeT extends StorageType>(obj: unknown, dataOrStorageType?: DataTypeT | StorageTypeT, storageType?: StorageTypeT): obj is PropertyModel<DataTypeT, StorageTypeT> {
    const dataType = isDataType(dataOrStorageType)
        ? dataOrStorageType
        : null;
    storageType = isStorageType(dataOrStorageType)
        ? dataOrStorageType
        : storageType;

    return isTypeStamped("property", obj)
        && isISequenced(obj)
        && isICreationTracked(obj)
        && "type" in obj
        && isPropertyTypeModel(obj.type, dataType, storageType)
        && "value" in obj
        && "units" in obj
        && isString(obj.units)
        && "updatedBy" in obj
        && isUserModel(obj.updatedBy)
        && "updatedOn" in obj
        && (isNullOrUndefined(obj.updatedOn)
            || isDate(obj.updatedOn));
}

export type SetPropertyInput<DataTypeT extends DataType = DataType, StorageTypeT extends StorageType = StorageType> = {
    type: PROPERTY_TYPE;
    value: DataTypeMap[DataTypeT][StorageTypeT];
    unitOfMeasure?: string;
    reference?: ENTITY;
}

export type PROPERTY = TypedNameOrId<"property">;
export function PROPERTY(idOrName: number | string): PROPERTY {
    return TypedNameOrId(idOrName, "property");
}

File: NETCore/Cedrus.Server/Controllers/V1/Models/PropertyModel.cs
using Juniper.Cedrus.Entities;
using Juniper.Units;

namespace Juniper.Cedrus.Controllers.V1;

public class PropertyModel : AbstractSequencedModel
{
    public string TypeStamp => "property";
    public EntityModel Entity { get; }
    public PropertyTypeModel Type { get; }
    public object Value { get; }
    public UnitOfMeasure Units { get; }
    public UserModel CreatedBy { get; }
    public DateTime CreatedOn { get; }
    public UserModel? UpdatedBy { get; }
    public DateTime? UpdatedOn { get; }
    public EntityModel? Reference { get; }

    public PropertyModel(Property property, Memoizer? memo = null)
        : base(property)
    {
        Type = property.Type.Memo(memo, () => new PropertyTypeModel(property.Type));
        Value = property.Value;
        Units = property.Units;
        CreatedBy = property.User.Memo(memo, () => new UserModel(property.User, memo));
        CreatedOn = property.CreatedOn;
        UpdatedBy = property.UpdatedByUser?.Memo(memo, () => new UserModel(property.UpdatedByUser, memo));
        UpdatedOn = property.UpdatedOn;
        Reference = property.Reference?.Memo(memo, () => new EntityModel(property.Reference, memo));
        Entity = property.Entity.Memo(memo, () => new EntityModel(property.Entity, memo));
    }
}
